clients/drcachesim/tracer/instr_counter.cpp

```diff
@@ -347,9 +425,157 @@ event_inscount_thread_init(void *drcontext)
         DELAY_COUNTDOWN_UNIT;
 }
 
+// Represents an interval as a <start,duration> pair in terms of number of instructions.
+struct instr_interval_t {
+    instr_interval_t(uint64 start, uint64 duration)
+        : start(start)
+        , duration(duration)
+    {
+    }
+
+    uint64 start;
+    uint64 duration;
+};
+
+// Function to order instruction intervals by start time in ascending order.
+static bool
+cmp_start_instr(const instr_interval_t &l, const instr_interval_t &r)
+{
+    return l.start < r.start;
+}
+
+static std::vector<instr_interval_t>
+parse_instr_intervals_file(std::string path_to_file)
+{
+    std::vector<instr_interval_t> instr_intervals;
+    std::ifstream file(path_to_file);
+    if (!file.is_open())
+        FATAL("Fatal error: failed to open file %s.\n", path_to_file.c_str());
+
+    std::string line;
+    while (std::getline(file, line)) {
+        // Ignore empty lines, if any.
+        if (line.empty())
+            continue;
+        std::stringstream ss(line);
+        std::string elem;
+        if (!std::getline(ss, elem, ','))
+            FATAL("Fatal error: start instruction not found.\n");
+        uint64 start = std::stoull(elem);
+        if (!std::getline(ss, elem, ','))
+            FATAL("Fatal error: instruction duration not found.\n");
+        uint64 duration = std::stoull(elem);
+        // Ignore the remaining comma-separated elements, if any.
+
+        instr_intervals.emplace_back(start, duration);
+    }
+    file.close();
+
+    if (instr_intervals.empty()) {
+        FATAL("Fatal error: -trace_instr_intervals_file %s contains no intervals.\n",
+              path_to_file.c_str());
+    }
+
+    // Enforcing constraints on intervals:
+    // 1) They need to be ordered by start time.
+    std::sort(instr_intervals.begin(), instr_intervals.end(), cmp_start_instr);
```

Comment:
Having the simpoints use-case in mind: the csv files we generate with the intervals might not be sorted, and since we plan to trace these irregular windows a bit before and after the actual given simpoints intervals they might also be overlapping. Having these constraint enforced here is more convenient than having to do it separately.

Is your worry about users providing a "malformed" csv file that is then transformed under their nose to comply with these constraints? Perhaps a warning would be sufficient?